bdk_electrum_example/src/main.rs

```diff
@@ -0,0 +1,277 @@
+            invalidate,
+            new_tip,
+        },
+    ))
+}
+
+pub fn perform_electrum_sync<K: Debug + Clone + Ord>(
+    client: &Client,
+    tracker: &mut KeychainTracker<K>,
+    chain: &mut SparseChain,
+    graph: &mut TxGraph,
+) -> anyhow::Result<()> {
+    let start = std::time::Instant::now();
+    let mut active_indexes = vec![];
+    for (keychain, descriptor) in tracker.iter_keychains(..) {
+        eprint!("scanning {:?} addresses indexes ", keychain);
+        let (last_active_index, checkpoint) = fetch_update_from_client(
+            graph,
+            &chain,
+            client,
+            &mut descriptor
+                .iter_all_scripts()
+                .enumerate()
+                .map(|(i, script)| (i as u32, script))
+                .inspect(|(i, _)| {
+                    use std::io::{self, Write};
+                    eprint!("{} ", i);
+                    let _ = io::stdout().flush();
+                }),
+            5,
+        )
+        .context("fetching transactions")?;
+
+        match chain.apply_update(checkpoint) {
+            Result::Ok(_) => eprintln!("success! ({}ms)", start.elapsed().as_millis()),
+            Result::Err(_reason) => {
+                unreachable!("we are the only ones accessing the tracker")
+            }
+        }
+
+        if let Some(last_active_index) = last_active_index {
+            active_indexes.push((keychain, last_active_index));
+        }
+    }
+
+    for (keychain, active_index) in active_indexes {
+        tracker.derive_spks(keychain, active_index);
+    }
+
+    tracker.sync(&graph);
+
+    Ok(())
+}
+
+#[derive(Debug, Clone, Copy, PartialOrd, Ord, PartialEq, Eq)]
+pub enum Keychain {
+    External,
+    Internal,
+}
+
+fn main() -> anyhow::Result<()> {
+    let args = App::parse();
+    let secp = Secp256k1::default();
+    let (descriptor, mut keymap) =
+        Descriptor::<DescriptorPublicKey>::parse_descriptor(&secp, &args.descriptor)?;
+
+    let mut tracker = KeychainTracker::default();
+    let mut sparse_chain = SparseChain::default();
+    let mut tx_graph = TxGraph::default();
+    tracker.add_keychain(Keychain::External, descriptor);
+
+    let internal = args
+        .change_descriptor
+        .map(|descriptor| Descriptor::<DescriptorPublicKey>::parse_descriptor(&secp, &descriptor))
+        .transpose()?;
+
+    let change_keychain = if let Some((internal_descriptor, internal_keymap)) = internal {
+        keymap.extend(internal_keymap);
+        tracker.add_keychain(Keychain::Internal, internal_descriptor);
+        Keychain::Internal
+    } else {
+        Keychain::External
+    };
+
+    // Construct a Electrum Client
+    use electrum_client::Config;
+    let config = Config::default();
+    let client = Client::from_config("ssl://electrum.blockstream.info:60002", config)?;
+    // Sync at boot
+    perform_electrum_sync(&client, &mut tracker, &mut sparse_chain, &mut tx_graph)?;
+
+    match args.command {
```

Comment:
Can't this whole thing just go in the example cli library like "process_cmd" and pass in the references above?

Reply:
Hmm ya.. Could be.. Makes sense to be there actually.. Don't remember why I put it here..

Reply:
I think I will understand it more once I understand the current core structures.. Seems a lot has changed..